### ISSUE TITLE
Misleading compilation error message in desugaring of closure with mutable reference

### ISSUE THREAD

**Issue by nirmangupta** [rust-lang/rust#72940]
I tried this code:

```rust
struct A(i32);

struct B {
    arr: Vec<A>
}

impl B {
    fn using_iterator(&mut self) -> Vec<&A> {
        (0..2).map(|i| &self.arr[i]).collect()
    }
}
```

I expected to see this happen:
While I understand that because of the mutable reference and how closures are desugared, the code is expected to not compile. But the error message seems misleading.
I expected to receive a compilation error message which explained multiple mutable borrows.
(discussion at - https://users.rust-lang.org/t/lifetimes-in-closures-with-captured-mutable-reference/43721 

Instead, this happened:
Received misleading error message 
```
9 |         (0..2).map(|i| &self.arr[i]).collect()
  |                    ^^^  ---- `self` is borrowed here
  |                    |
  |                    may outlive borrowed value `self`
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018

### Meta
Tried on both stable and nightly
`rustc --version --verbose`:
```
binary: rustc
commit-hash: 4fb7144ed159f94491249e86d5bbd033b5d60550
commit-date: 2020-04-20
host: x86_64-apple-darwin
release: 1.43.0
LLVM version: 9.0
```

```
rustc 1.45.0-nightly (fe10f1a49 2020-06-02)
binary: rustc
commit-hash: fe10f1a49f5ca46e57261b95f46f519523f418fe
commit-date: 2020-06-02
host: x86_64-apple-darwin
release: 1.45.0-nightly
LLVM version: 10.0
```
Backtrace:

**Comment by estebank** [comment]
Current output:

```
error[E0373]: closure may outlive the current function, but it borrows `self`, which is owned by the current function
 --> src/main.rs:9:20
  |
9 |         (0..2).map(|i| &self.arr[i]).collect()
  |                    ^^^  -------- `self` is borrowed here
  |                    |
  |                    may outlive borrowed value `self`
  |
note: closure is returned here
 --> src/main.rs:9:9
  |
9 |         (0..2).map(|i| &self.arr[i]).collect()
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: to force the closure to take ownership of `self` (and any other referenced variables), use the `move` keyword
  |
9 |         (0..2).map(move |i| &self.arr[i]).collect()
  |                    ++++
```

If applied, you get

```
error: captured variable cannot escape `FnMut` closure body
 --> src/main.rs:9:28
  |
9 |         (0..2).map(move|i| &self.arr[i]).collect()
  |                          - ^^^^^^^^^^^^ returns a reference to a captured variable which escapes the closure body
  |                          |
  |                          inferred to be a `FnMut` closure
  |
  = note: `FnMut` closures only have access to their captured variables while they are executing...
  = note: ...therefore, they cannot allow references to captured variables to escape
```
